// nativescript-telerik-ui-pro/gauges/initializers/gauges-initializers.js
var RadGaugeValueMapper = (function () {
    function RadGaugeValueMapper() {
    }
    RadGaugeValueMapper.prototype.onTitleChanged = function (eventData, gauge) {
    };
    RadGaugeValueMapper.prototype.onSubtitleChanged = function (eventData, gauge) {
    };
    RadGaugeValueMapper.prototype.onTitleStyleChanged = function (eventData, gauge) {
    };
    RadGaugeValueMapper.prototype.onSubtitleStyleChanged = function (eventData, gauge) {
    };
    RadGaugeValueMapper.prototype.updateNativeObject = function (gauge) {
    };
    return RadGaugeValueMapper;
}());
exports.RadGaugeValueMapper = RadGaugeValueMapper;
var GaugeScaleValueMapper = (function () {
    function GaugeScaleValueMapper() {
    }
    GaugeScaleValueMapper.prototype.onMinimumChanged = function (eventData, scale) {
    };
    GaugeScaleValueMapper.prototype.onMaximumChanged = function (eventData, scale) {
    };
    GaugeScaleValueMapper.prototype.onIndicatorsChange = function (eventData, scale) {
    };
    GaugeScaleValueMapper.prototype.onScaleStyleChanged = function (eventData, scale) {
    };
    GaugeScaleValueMapper.prototype.reloadIndicators = function (scale) {
    };
    GaugeScaleValueMapper.prototype.updateNativeObject = function (scale) {
    };
    return GaugeScaleValueMapper;
}());
exports.GaugeScaleValueMapper = GaugeScaleValueMapper;
var NeedleIndicatorValueMapper = (function () {
    function NeedleIndicatorValueMapper() {
    }
    NeedleIndicatorValueMapper.prototype.onIsAnimatedChanged = function (eventData, indicator) {
    };
    NeedleIndicatorValueMapper.prototype.onAnimationDurationChanged = function (eventData, indicator) {
    };
    NeedleIndicatorValueMapper.prototype.onNeedleStyleChanged = function (eventData, needle) {
    };
    NeedleIndicatorValueMapper.prototype.updateNativeObject = function (needle) {
    };
    return NeedleIndicatorValueMapper;
}());
exports.NeedleIndicatorValueMapper = NeedleIndicatorValueMapper;
var BarIndicatorValueMapper = (function () {
    function BarIndicatorValueMapper() {
    }
    BarIndicatorValueMapper.prototype.onIsAnimatedChanged = function (eventData, indicator) {
    };
    BarIndicatorValueMapper.prototype.onAnimationDurationChanged = function (eventData, indicator) {
    };
    BarIndicatorValueMapper.prototype.onMinimumValueChanged = function (eventData, indicator) {
    };
    BarIndicatorValueMapper.prototype.onMaximumValueChanged = function (eventData, indicator) {
    };
    BarIndicatorValueMapper.prototype.onBarStyleChanged = function (eventData, indicator) {
    };
    BarIndicatorValueMapper.prototype.onLocationChanged = function (eventData, indicator) {
    };
    BarIndicatorValueMapper.prototype.updateNativeObject = function (barIndicator) {
    };
    return BarIndicatorValueMapper;
}());
exports.BarIndicatorValueMapper = BarIndicatorValueMapper;
