// nativescript-telerik-ui-pro/chart/initializers/chart-initializers.js
var ChartBaseValueMapper = (function () {
    function ChartBaseValueMapper() {
    }
    ChartBaseValueMapper.prototype.onLegendChanged = function (data, chart) { };
    ChartBaseValueMapper.prototype.onPalettesChanged = function (data, chart) { };
    ChartBaseValueMapper.prototype.onSeriesChanged = function (data, chart) { };
    ChartBaseValueMapper.prototype.onAnnotationsChanged = function (data, chart) { };
    ChartBaseValueMapper.prototype.onSelectionModeChanged = function (data, chart) { };
    ChartBaseValueMapper.prototype.loadSeries = function (chart) { };
    ChartBaseValueMapper.prototype.loadAnnotations = function (chart) { };
    ChartBaseValueMapper.prototype.reloadPalettes = function (chart) { };
    ChartBaseValueMapper.prototype.updateHorizontalAxisPalette = function (chart) { };
    ChartBaseValueMapper.prototype.updateVerticalAxisPalette = function (chart) { };
    return ChartBaseValueMapper;
}());
exports.ChartBaseValueMapper = ChartBaseValueMapper;
var SplineSeriesValueMapper = (function () {
    function SplineSeriesValueMapper() {
    }
    SplineSeriesValueMapper.prototype.onCategoryPropertyChanged = function (data, series) { };
    SplineSeriesValueMapper.prototype.onStackModePropertyChanged = function (data, series) { };
    SplineSeriesValueMapper.prototype.updateNative = function (series) { };
    SplineSeriesValueMapper.prototype.onHorizontalAxisChanged = function (data, series) { };
    SplineSeriesValueMapper.prototype.onVerticalAxisChanged = function (data, series) { };
    SplineSeriesValueMapper.prototype.onLegendTitleChanged = function (data, series) { };
    SplineSeriesValueMapper.prototype.onItemsChanged = function (data, series) { };
    SplineSeriesValueMapper.prototype.onValuePropertyChanged = function (data, series) { };
    SplineSeriesValueMapper.prototype.onShowLabelsChanged = function (data, series) { };
    SplineSeriesValueMapper.prototype.onLabelStyleChanged = function (data, series) { };
    SplineSeriesValueMapper.prototype.onSelectionModeChanged = function (data, series) { };
    SplineSeriesValueMapper.prototype.applyLabelStyle = function (series, chart) { };
    return SplineSeriesValueMapper;
}());
exports.SplineSeriesValueMapper = SplineSeriesValueMapper;
var SplineAreaSeriesValueMapper = (function (_super) {
    __extends(SplineAreaSeriesValueMapper, _super);
    function SplineAreaSeriesValueMapper() {
        _super.apply(this, arguments);
    }
    return SplineAreaSeriesValueMapper;
}(SplineSeriesValueMapper));
exports.SplineAreaSeriesValueMapper = SplineAreaSeriesValueMapper;
var ChartSeriesValueMapper = (function () {
    function ChartSeriesValueMapper() {
    }
    ChartSeriesValueMapper.prototype.onLegendTitleChanged = function (data, series) { };
    ChartSeriesValueMapper.prototype.onItemsChanged = function (data, series) { };
    ChartSeriesValueMapper.prototype.onValuePropertyChanged = function (data, series) { };
    ChartSeriesValueMapper.prototype.onShowLabelsChanged = function (data, series) { };
    ChartSeriesValueMapper.prototype.onLabelStyleChanged = function (data, series) { };
    ChartSeriesValueMapper.prototype.onSelectionModeChanged = function (data, series) { };
    ChartSeriesValueMapper.prototype.applyLabelStyle = function (series, chart) { };
    return ChartSeriesValueMapper;
}());
exports.ChartSeriesValueMapper = ChartSeriesValueMapper;
var CartesianSeriesValueMapper = (function (_super) {
    __extends(CartesianSeriesValueMapper, _super);
    function CartesianSeriesValueMapper() {
        _super.apply(this, arguments);
    }
    CartesianSeriesValueMapper.prototype.onHorizontalAxisChanged = function (data, series) { };
    CartesianSeriesValueMapper.prototype.onVerticalAxisChanged = function (data, series) { };
    return CartesianSeriesValueMapper;
}(ChartSeriesValueMapper));
exports.CartesianSeriesValueMapper = CartesianSeriesValueMapper;
var CategoricalSeriesValueMapper = (function (_super) {
    __extends(CategoricalSeriesValueMapper, _super);
    function CategoricalSeriesValueMapper() {
        _super.apply(this, arguments);
    }
    CategoricalSeriesValueMapper.prototype.onCategoryPropertyChanged = function (data, series) { };
    CategoricalSeriesValueMapper.prototype.onStackModePropertyChanged = function (data, series) { };
    CategoricalSeriesValueMapper.prototype.updateNative = function (series) { };
    return CategoricalSeriesValueMapper;
}(CartesianSeriesValueMapper));
exports.CategoricalSeriesValueMapper = CategoricalSeriesValueMapper;
var PieSeriesValueMapper = (function (_super) {
    __extends(PieSeriesValueMapper, _super);
    function PieSeriesValueMapper() {
        _super.apply(this, arguments);
    }
    PieSeriesValueMapper.prototype.onLabelPropertyChanged = function (data, series) { };
    PieSeriesValueMapper.prototype.onExpandRadiusChanged = function (data, series) { };
    PieSeriesValueMapper.prototype.onOuterRadiusFactorChanged = function (data, series) { };
    PieSeriesValueMapper.prototype.onStartAngleChanged = function (data, series) { };
    PieSeriesValueMapper.prototype.onEndAngleChanged = function (data, series) { };
    PieSeriesValueMapper.prototype.onShowPercentageChanged = function (data, series) { };
    return PieSeriesValueMapper;
}(ChartSeriesValueMapper));
exports.PieSeriesValueMapper = PieSeriesValueMapper;
var DonutSeriesValueMapper = (function (_super) {
    __extends(DonutSeriesValueMapper, _super);
    function DonutSeriesValueMapper() {
        _super.apply(this, arguments);
    }
    DonutSeriesValueMapper.prototype.onInnerRadiusFactorChanged = function (data, series) { };
    return DonutSeriesValueMapper;
}(PieSeriesValueMapper));
exports.DonutSeriesValueMapper = DonutSeriesValueMapper;
var BarSeriesValueMapper = (function (_super) {
    __extends(BarSeriesValueMapper, _super);
    function BarSeriesValueMapper() {
        _super.apply(this, arguments);
    }
    return BarSeriesValueMapper;
}(CategoricalSeriesValueMapper));
exports.BarSeriesValueMapper = BarSeriesValueMapper;
var RangeBarSeriesValueMapper = (function (_super) {
    __extends(RangeBarSeriesValueMapper, _super);
    function RangeBarSeriesValueMapper() {
        _super.apply(this, arguments);
    }
    RangeBarSeriesValueMapper.prototype.onHighPropertyNameChanged = function (data, series) { };
    RangeBarSeriesValueMapper.prototype.onLowPropertyNameChanged = function (data, series) { };
    return RangeBarSeriesValueMapper;
}(CategoricalSeriesValueMapper));
exports.RangeBarSeriesValueMapper = RangeBarSeriesValueMapper;
var LineSeriesValueMapper = (function (_super) {
    __extends(LineSeriesValueMapper, _super);
    function LineSeriesValueMapper() {
        _super.apply(this, arguments);
    }
    return LineSeriesValueMapper;
}(CategoricalSeriesValueMapper));
exports.LineSeriesValueMapper = LineSeriesValueMapper;
var BubbleSeriesValueMapper = (function (_super) {
    __extends(BubbleSeriesValueMapper, _super);
    function BubbleSeriesValueMapper() {
        _super.apply(this, arguments);
    }
    BubbleSeriesValueMapper.prototype.onBubbleSizePropertyChanged = function (data, series) { };
    BubbleSeriesValueMapper.prototype.onBubbleScalePropertyChanged = function (data, series) { };
    return BubbleSeriesValueMapper;
}(CategoricalSeriesValueMapper));
exports.BubbleSeriesValueMapper = BubbleSeriesValueMapper;
var CartesianAxisValueMapper = (function () {
    function CartesianAxisValueMapper() {
    }
    CartesianAxisValueMapper.prototype.onLineHiddenChanged = function (data, axis) { };
    CartesianAxisValueMapper.prototype.onLineThicknessChanged = function (data, axis) { };
    CartesianAxisValueMapper.prototype.onLineColorChanged = function (data, axis) { };
    CartesianAxisValueMapper.prototype.onLabelTextColorChanged = function (data, axis) { };
    CartesianAxisValueMapper.prototype.onLabelMarginChanged = function (data, axis) { };
    CartesianAxisValueMapper.prototype.onLabelRotationAngleChanged = function (data, axis) { };
    CartesianAxisValueMapper.prototype.onLabelFitModeChanged = function (data, axis) { };
    CartesianAxisValueMapper.prototype.onLabelLayoutModeChanged = function (data, axis) { };
    CartesianAxisValueMapper.prototype.onLabelFormatChanged = function (data, axis) { };
    CartesianAxisValueMapper.prototype.onHorizontalLocationChanged = function (data, axis) { };
    CartesianAxisValueMapper.prototype.onVerticalLocationChanged = function (data, axis) { };
    CartesianAxisValueMapper.prototype.onLabelSizeChanged = function (data, axis) { };
    CartesianAxisValueMapper.prototype.onAllowZoomChanged = function (data, axis) { };
    CartesianAxisValueMapper.prototype.onAllowPanChanged = function (data, axis) { };
    CartesianAxisValueMapper.prototype.onHiddenChanged = function (data, axis) { };
    return CartesianAxisValueMapper;
}());
exports.CartesianAxisValueMapper = CartesianAxisValueMapper;
var AreaSeriesValueMapper = (function (_super) {
    __extends(AreaSeriesValueMapper, _super);
    function AreaSeriesValueMapper() {
        _super.apply(this, arguments);
    }
    return AreaSeriesValueMapper;
}(LineSeriesValueMapper));
exports.AreaSeriesValueMapper = AreaSeriesValueMapper;
var CategoricalAxisValueMapper = (function (_super) {
    __extends(CategoricalAxisValueMapper, _super);
    function CategoricalAxisValueMapper() {
        _super.apply(this, arguments);
    }
    CategoricalAxisValueMapper.prototype.onMajorTickIntervalChanged = function (data, axis) { };
    CategoricalAxisValueMapper.prototype.onPlotModeChanged = function (data, axis) { };
    return CategoricalAxisValueMapper;
}(CartesianAxisValueMapper));
exports.CategoricalAxisValueMapper = CategoricalAxisValueMapper;
var DateTimeAxisValueMapper = (function (_super) {
    __extends(DateTimeAxisValueMapper, _super);
    function DateTimeAxisValueMapper() {
        _super.apply(this, arguments);
    }
    DateTimeAxisValueMapper.prototype.onDateTimeComponentChanged = function (data, axis) { };
    DateTimeAxisValueMapper.prototype.onDateFormatChanged = function (data, axis) { };
    DateTimeAxisValueMapper.prototype.onSourceDateFormatChanged = function (data, axis) { };
    return DateTimeAxisValueMapper;
}(CategoricalAxisValueMapper));
exports.DateTimeAxisValueMapper = DateTimeAxisValueMapper;
var OhlcSeriesValueMapper = (function (_super) {
    __extends(OhlcSeriesValueMapper, _super);
    function OhlcSeriesValueMapper() {
        _super.apply(this, arguments);
    }
    OhlcSeriesValueMapper.prototype.onHighPropertyNameChanged = function (data, series) { };
    OhlcSeriesValueMapper.prototype.onLowPropertyNameChanged = function (data, series) { };
    OhlcSeriesValueMapper.prototype.onOpenPropertyNameChanged = function (data, series) { };
    OhlcSeriesValueMapper.prototype.onClosePropertyNameChanged = function (data, series) { };
    return OhlcSeriesValueMapper;
}(CategoricalSeriesValueMapper));
exports.OhlcSeriesValueMapper = OhlcSeriesValueMapper;
var CandlestickSeriesValueMapper = (function (_super) {
    __extends(CandlestickSeriesValueMapper, _super);
    function CandlestickSeriesValueMapper() {
        _super.apply(this, arguments);
    }
    return CandlestickSeriesValueMapper;
}(OhlcSeriesValueMapper));
exports.CandlestickSeriesValueMapper = CandlestickSeriesValueMapper;
var ScatterSeriesValueMapper = (function (_super) {
    __extends(ScatterSeriesValueMapper, _super);
    function ScatterSeriesValueMapper() {
        _super.apply(this, arguments);
    }
    ScatterSeriesValueMapper.prototype.onXPropertyChanged = function (data, series) { };
    ScatterSeriesValueMapper.prototype.onYPropertyChanged = function (data, series) { };
    return ScatterSeriesValueMapper;
}(CartesianSeriesValueMapper));
exports.ScatterSeriesValueMapper = ScatterSeriesValueMapper;
var ScatterBubbleSeriesValueMapper = (function (_super) {
    __extends(ScatterBubbleSeriesValueMapper, _super);
    function ScatterBubbleSeriesValueMapper() {
        _super.apply(this, arguments);
    }
    ScatterBubbleSeriesValueMapper.prototype.onBubbleSizePropertyChanged = function (data, series) {
    };
    ScatterBubbleSeriesValueMapper.prototype.onBubbleScaleChanged = function (data, series) { };
    return ScatterBubbleSeriesValueMapper;
}(ScatterSeriesValueMapper));
exports.ScatterBubbleSeriesValueMapper = ScatterBubbleSeriesValueMapper;
