// nativescript-telerik-ui-pro/chart/visualization/views/chart-legend-view.js
var legendViewCommonModule = require('./chart-legend-view-common');
// NOTE: This is a dummy file used to fix Typescript errors while developing.
// The NativeScript build overwrites it with one of the corresponding .ios.ts or .android.ts
// files from the same folder.
var RadLegendView = (function (_super) {
    __extends(RadLegendView, _super);
    function RadLegendView() {
        _super.apply(this, arguments);
    }
    RadLegendView.prototype.onPositionChanged = function (data) {
    };
    RadLegendView.prototype.onVerticalOffsetChanged = function (data) {
    };
    RadLegendView.prototype.onHorizontalOffsetChanged = function (data) {
    };
    RadLegendView.prototype.onOffsetOriginChanged = function (data) {
    };
    RadLegendView.prototype.onTitleChanged = function (data) {
    };
    RadLegendView.prototype.updateLegendPosition = function (chartView) {
    };
    RadLegendView.prototype.updateLegendView = function (chartView) {
    };
    return RadLegendView;
}(legendViewCommonModule.RadLegendView));
exports.RadLegendView = RadLegendView;
